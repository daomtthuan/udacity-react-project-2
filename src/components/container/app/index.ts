import AppContainer from './_.view';

export default AppContainer;
