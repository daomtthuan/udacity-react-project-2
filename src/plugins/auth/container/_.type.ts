import { PropsWithChildren } from 'react';

export type AuthContainerProps = PropsWithChildren;
