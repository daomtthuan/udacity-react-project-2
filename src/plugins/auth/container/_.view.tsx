import { Location, Navigate, useLocation } from 'react-router-dom';
import { useAppSelector } from '~plugins/store';

import { AuthContainerProps } from './_.type';

const pathConfig = {
  signIn: '/',
  home: '/home',

  guest: ['/', '/sign-up'],
};

export default function AuthContainer({ children }: Readonly<AuthContainerProps>) {
  const location: Location<string | null> = useLocation();

  const authUser = useAppSelector((state) => state.auth.user);

  // Not authenticated
  if (!authUser) {
    // Location is not guest routes
    if (!pathConfig.guest.includes(location.pathname)) {
      return <Navigate to={pathConfig.signIn} state={location.pathname} />;
    }

    return children;
  }

  // Authenticated
  else {
    // Location has state
    if (location.state) {
      return <Navigate to={location.state} state={null} />;
    }

    // Location is guest routes
    if (pathConfig.guest.includes(location.pathname)) {
      return <Navigate to={pathConfig.home} state={null} />;
    }

    return children;
  }
}
