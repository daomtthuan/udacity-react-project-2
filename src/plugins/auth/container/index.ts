import AuthContainer from './_.view';

export default AuthContainer;
