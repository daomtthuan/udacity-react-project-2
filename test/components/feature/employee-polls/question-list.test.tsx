import QuestionList, { QuestionListProps } from '~components/feature/employee-polls/question-list';
import { TestContainer } from '~test/mocks/container';

import { render } from '@testing-library/react';

describe('QuestionList renders correctly', () => {
  test.each<QuestionListProps>([
    {
      title: 'Unanswered',
      items: [
        {
          id: 'loxhs1bqm25b708cmbf3g',
          author: 'tylermcginnis',
          timestamp: 1482579767190,
          optionOne: {
            votes: [],
            text: 'have code reviews conducted by peers',
          },
          optionTwo: {
            votes: ['sarahedo'],
            text: 'have code reviews conducted by managers',
          },
          authorUser: {
            id: 'tylermcginnis',
            password: 'abc321',
            name: 'Tyler McGinnis',
            avatarURL: '/images/avatar/2.png',
            answers: {
              vthrdm985a262al8qx3do: 'optionOne',
              xj352vofupe1dqz9emx13r: 'optionTwo',
            },
            questions: ['loxhs1bqm25b708cmbf3g', 'vthrdm985a262al8qx3do'],
          },
        },
        {
          id: 'a0xo5jhg6y7tj2hwt3t6k',
          timestamp: 1709492013513,
          author: 'thuandmt',
          optionOne: {
            votes: [],
            text: 'Helllo',
          },
          optionTwo: {
            votes: ['thuandmt'],
            text: 'AAA',
          },
          authorUser: {
            id: 'thuandmt',
            password: '123',
            name: 'Thuan Dao',
            avatarURL: '/images/avatar/7.png',
            answers: {
              'xj352vofupe1dqz9emx13r': 'optionOne',
              'vthrdm985a262al8qx3do': 'optionTwo',
              'am8ehyc8byjqgar0jgpub9': 'optionOne',
              '6ni6ok3ym7mf1p33lnez': 'optionTwo',
              '8xf0y6ziyjabvozdd253nd': 'optionOne',
              'a0xo5jhg6y7tj2hwt3t6k': 'optionTwo',
            },
            questions: [],
          },
        },
      ],
      buttonText: 'Answer',
    },

    {
      title: 'Answered',
      items: [
        {
          id: 'a0xo5jhg6y7tj2hwt3t6k',
          timestamp: 1709492013513,
          author: 'thuandmt',
          optionOne: {
            votes: [],
            text: 'Helllo',
          },
          optionTwo: {
            votes: ['thuandmt'],
            text: 'AAA',
          },
          authorUser: {
            id: 'thuandmt',
            password: '123',
            name: 'Thuan Dao',
            avatarURL: '/images/avatar/7.png',
            answers: {
              'xj352vofupe1dqz9emx13r': 'optionOne',
              'vthrdm985a262al8qx3do': 'optionTwo',
              'am8ehyc8byjqgar0jgpub9': 'optionOne',
              '6ni6ok3ym7mf1p33lnez': 'optionTwo',
              '8xf0y6ziyjabvozdd253nd': 'optionOne',
              'a0xo5jhg6y7tj2hwt3t6k': 'optionTwo',
            },
            questions: [],
          },
        },
        {
          id: 'xj352vofupe1dqz9emx13r',
          author: 'mtsamis',
          timestamp: 1493579767190,
          optionOne: {
            votes: ['mtsamis', 'thuandmt'],
            text: 'deploy to production once every two weeks',
          },
          optionTwo: {
            votes: ['tylermcginnis'],
            text: 'deploy to production once every month',
          },
          authorUser: {
            id: 'mtsamis',
            password: 'xyz123',
            name: 'Mike Tsamis',
            avatarURL: '/images/avatar/3.png',
            answers: {
              'xj352vofupe1dqz9emx13r': 'optionOne',
              'vthrdm985a262al8qx3do': 'optionTwo',
              '6ni6ok3ym7mf1p33lnez': 'optionOne',
            },
            questions: ['6ni6ok3ym7mf1p33lnez', 'xj352vofupe1dqz9emx13r'],
          },
        },
        {
          id: 'vthrdm985a262al8qx3do',
          author: 'tylermcginnis',
          timestamp: 1489579767190,
          optionOne: {
            votes: ['tylermcginnis'],
            text: 'take a course on ReactJS',
          },
          optionTwo: {
            votes: ['mtsamis', 'thuandmt'],
            text: 'take a course on unit testing with Jest',
          },
          authorUser: {
            id: 'tylermcginnis',
            password: 'abc321',
            name: 'Tyler McGinnis',
            avatarURL: '/images/avatar/2.png',
            answers: {
              vthrdm985a262al8qx3do: 'optionOne',
              xj352vofupe1dqz9emx13r: 'optionTwo',
            },
            questions: ['loxhs1bqm25b708cmbf3g', 'vthrdm985a262al8qx3do'],
          },
        },
      ],
      buttonText: 'Result',
    },
  ])('QuestionList render as expected with props %o', (props) => {
    const component = render(
      <TestContainer>
        <QuestionList {...props} />
      </TestContainer>,
    );

    expect(component.baseElement).toMatchSnapshot();
  });
});
