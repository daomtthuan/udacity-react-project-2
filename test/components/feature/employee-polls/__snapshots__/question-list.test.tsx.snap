// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly QuestionList should render as expected 1`] = `
<body>
  <div>
    <div
      class="border d-flex flex-column rounded shadow-sm"
    >
      <div
        class="border-bottom bg-light p-3 text-center"
      >
        <h2
          class="text-primary mb-0"
        >
          Unanswered
        </h2>
      </div>
      <div
        class="bg-white p-3"
      >
        <div
          class="row row-cols-1 row-cols-md-3 g-3"
        >
          <div
            class="col"
          >
            <div
              class="card shadow-sm h-100"
            >
              <div
                class="card-body text-center"
              >
                <div
                  class="card-title"
                >
                  <span
                    class="fs-7 text-secondary"
                  >
                    Poll by
                  </span>
                  <h5>
                    Tyler McGinnis
                  </h5>
                </div>
                <p
                  class="card-text text-secondary"
                >
                  6:42:47 PM - 12/24/2016
                </p>
              </div>
              <div
                class="card-footer"
              >
                <a
                  class="btn btn-outline-primary w-100"
                  href="/questions/loxhs1bqm25b708cmbf3g"
                >
                  Answer
                </a>
              </div>
            </div>
          </div>
          <div
            class="col"
          >
            <div
              class="card shadow-sm h-100"
            >
              <div
                class="card-body text-center"
              >
                <div
                  class="card-title"
                >
                  <span
                    class="fs-7 text-secondary"
                  >
                    Poll by
                  </span>
                  <h5>
                    Thuan Dao
                  </h5>
                </div>
                <p
                  class="card-text text-secondary"
                >
                  1:53:33 AM - 3/4/2024
                </p>
              </div>
              <div
                class="card-footer"
              >
                <a
                  class="btn btn-outline-primary w-100"
                  href="/questions/a0xo5jhg6y7tj2hwt3t6k"
                >
                  Answer
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Renders correctly QuestionList should render as expected 2`] = `
<body>
  <div>
    <div
      class="border d-flex flex-column rounded shadow-sm"
    >
      <div
        class="border-bottom bg-light p-3 text-center"
      >
        <h2
          class="text-primary mb-0"
        >
          Answered
        </h2>
      </div>
      <div
        class="bg-white p-3"
      >
        <div
          class="row row-cols-1 row-cols-md-3 g-3"
        >
          <div
            class="col"
          >
            <div
              class="card shadow-sm h-100"
            >
              <div
                class="card-body text-center"
              >
                <div
                  class="card-title"
                >
                  <span
                    class="fs-7 text-secondary"
                  >
                    Poll by
                  </span>
                  <h5>
                    Thuan Dao
                  </h5>
                </div>
                <p
                  class="card-text text-secondary"
                >
                  1:53:33 AM - 3/4/2024
                </p>
              </div>
              <div
                class="card-footer"
              >
                <a
                  class="btn btn-outline-primary w-100"
                  href="/questions/a0xo5jhg6y7tj2hwt3t6k"
                >
                  Result
                </a>
              </div>
            </div>
          </div>
          <div
            class="col"
          >
            <div
              class="card shadow-sm h-100"
            >
              <div
                class="card-body text-center"
              >
                <div
                  class="card-title"
                >
                  <span
                    class="fs-7 text-secondary"
                  >
                    Poll by
                  </span>
                  <h5>
                    Mike Tsamis
                  </h5>
                </div>
                <p
                  class="card-text text-secondary"
                >
                  2:16:7 AM - 5/1/2017
                </p>
              </div>
              <div
                class="card-footer"
              >
                <a
                  class="btn btn-outline-primary w-100"
                  href="/questions/xj352vofupe1dqz9emx13r"
                >
                  Result
                </a>
              </div>
            </div>
          </div>
          <div
            class="col"
          >
            <div
              class="card shadow-sm h-100"
            >
              <div
                class="card-body text-center"
              >
                <div
                  class="card-title"
                >
                  <span
                    class="fs-7 text-secondary"
                  >
                    Poll by
                  </span>
                  <h5>
                    Tyler McGinnis
                  </h5>
                </div>
                <p
                  class="card-text text-secondary"
                >
                  7:9:27 PM - 3/15/2017
                </p>
              </div>
              <div
                class="card-footer"
              >
                <a
                  class="btn btn-outline-primary w-100"
                  href="/questions/vthrdm985a262al8qx3do"
                >
                  Result
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
